### PR TITLE
Make GM behave as DM

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Make GM behave as DM

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-10109
Fix https://github.com/mattermost/mattermost/issues/19998

#### Release Note
```release-note
Make GMs behave like DMs, showing a new mention on every message.
```

### CODE REVIEW

**Comment by johndavidlugtu** [comment]
Good stuff @larkox 

The default desktop and mobile notification preference should be "For all activity" for Group Messages

Screenshot 2023-08-17 at 6 20 53 PM

Also, to reiterate @marianunez point [here](https://community.mattermost.com/core/pl/4ib1sm473fg38dxbu7sy8xfiwe), the difference between `For all activity` and `Only for mentions` is that:

- `For all activity` will generate an audio + OS notification + increment the badge count for ALL messages
- `Only for mentions` will generate an audio + OS notification + increment the badge count only for messages that a user is explicitly @ mentioned

**Comment by larkox** [comment]
@marianunez @abhijit-singh The latest test server should have all the relevant changes, so feel free to test them. In the meanwhile, I will start adding people for the dev reviews.

**Comment by marianunez** [comment]
> @marianunez @abhijit-singh The latest test server should have all the relevant changes, so feel free to test them. In the meanwhile, I will start adding people for the dev reviews.

@larkox for the existing GMs notice, should the primary button be `Dismiss` as originally designed? What is the behavior of `Remind me Later`?

Screen Shot 2023-08-23 at 9 01 21 AM

**Comment by larkox** [comment]
@marianunez Good question. Before, in system notices, no button had an emphasis, but were located in that order (first remind me later, then don't show again). I just applied the primary style to the one on the left, as the design, but didn't think about the content of the buttons.

Making a change here would imply making a change to all notices if we want to keep things consistent. That, or add an exception for this one, but I will leave that to UX. Exceptions are not great for keeping things consistent :P

Regarding the behavior...

Remind me later will store in the "local state" that you have seen that notice. When the local state is purged (mainly when you refresh the page in a browser), that value disappears, so this notice shows again.

On the contrary, if you click `Don't show again`, it will send a message to the server to update the server state (and local state while at it). Therefore, you will never see that notice again for any other client or session.

**Comment by marianunez** [comment]
> @marianunez Good question. Before, in system notices, no button had an emphasis, but were located in that order (first remind me later, then don't show again). I just applied the primary style to the one on the left, as the design, but didn't think about the content of the buttons.
> 
> Making a change here would imply making a change to all notices if we want to keep things consistent. That, or add an exception for this one, but I will leave that to UX. Exceptions are not great for keeping things consistent :P
> 
> Regarding the behavior...
> 
> Remind me later will store in the "local state" that you have seen that notice. When the local state is purged (mainly when you refresh the page in a browser), that value disappears, so this notice shows again.
> 
> On the contrary, if you click `Don't show again`, it will send a message to the server to update the server state (and local state while at it). Therefore, you will never see that notice again for any other client or session.

I see, I'm all for reusing the notice component here so these options sound good to me. Will defer to @abhijit-singh on the UX details for the action buttons.

**Review by marianunez — COMMENTED** [review]
Overall looks good, would be good to add a bit more test coverage, ie. on the changes in the notifications settings screen.  Also, not sure if the mobile notifications changes you want to include in a separate PR?

**Review by harshilsharma63 — APPROVED** [review]
👌👌👌

File: webapp/channels/src/components/system_notice/notices.tsx [hunk 1]
```diff
@@ -31,12 +31,11 @@ const notices: Notice[] = [
         name: 'apiv3_deprecation',
         adminOnly: true,
         title: (
-            <FormattedMarkdownMessage
+            <FormattedMessage
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/system_notice/notices.tsx:34]: 👍🏽 for this changes

**Review by M-ZubairAhmed — APPROVED** [review]

**Review by abhijit-singh — APPROVED** [review]
Looks good @larkox! Thank you!

Another aspect of this was changing the text in the `Notification settings` from `Only for mentions and direct messages` to  `Only for mentions, direct messages, and group messages` under both Desktop and Mobile notifications sections. It doesn't have to be done in this PR though.
image

**Comment by larkox** [comment]
@abhijit-singh Forgot about that one :bow: Will do it in this PR.

File: webapp/channels/src/components/channel_notifications_modal/__snapshots__/channel_notifications_modal.test.tsx.snap [hunk 2]
```diff
@@ -137,6 +142,7 @@ exports[`components/channel_notifications_modal/ChannelNotificationsModal should
             channelAutoFollowThreads="off"
             expand={false}
             ignoreChannelMentions="off"
+            isGM={false}
```
Comment (marianunez) [inline webapp/channels/src/components/channel_notifications_modal/__snapshots__/channel_notifications_modal.test.tsx.snap:145]: Should we add snapshots for GM `true` case?

**Review by marianunez — APPROVED** [review]
Are there any other opportunities to add more tests in the webapp?

Also, adding @yasserfaraazkhan for `QA Review`

**Review by yasserfaraazkhan — COMMENTED** [review]
HI @larkox @johndavidlugtu 
I've tested this in mobile, desktop and web browser. 
The badge count, notification for new messages are functioning as expected.

I had one scenario to check,
if the UserA -> sends a message to UserB, but deletes it before UserB could read it, what should be the behaviour ?
currently the Badge count persists conveying there's a unread message(s). This is same for DM & GM.
would this be expected behaviour ?

**Review by yasserfaraazkhan — APPROVED** [review]

File: webapp/channels/src/components/channel_notifications_modal/__snapshots__/channel_notifications_modal.test.tsx.snap (hunk 2 above)
Comment (larkox) [reply webapp/channels/src/components/channel_notifications_modal/__snapshots__/channel_notifications_modal.test.tsx.snap:145]: I added several more tests all around the PR. I haven't added any new e2e tests, but I think they are not needed.

Either way, if you think there is any particular behavior that has to be added to the e2e tests, let me know, and I will look into it.

Comment (marianunez) [reply webapp/channels/src/components/channel_notifications_modal/__snapshots__/channel_notifications_modal.test.tsx.snap:145]: LGTM - was looking for covering the case for the settings change in the GM which I see you've covered. Also, thanks for the additional mentions tests in the server 💯

**Review by johndavidlugtu — APPROVED** [review]
Great stuff @larkox

**Comment by johndavidlugtu** [comment]
Hey @amyblais can we get this for v9.1 along with the [ability to convert GM to private channel](https://github.com/mattermost/mattermost/pull/24421#issuecomment-1725442199)?

**Pull request merged**